aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -576,55 +573,27 @@ Tensor& NestedTensor_mul__Tensor(Tensor& self, const Tensor& other) {
 }
 
 Tensor select_nested(const Tensor& self, int64_t dim, int64_t index) {
+  auto self_ptr = get_nested_tensor_impl(self);
+  int64_t positive_dim = at::maybe_wrap_dim(dim, self_ptr->dim());
   TORCH_CHECK(
-    dim == 0,
+    positive_dim == 0,
     "NestedTensor can only be selected along dimension 0 ",
     "got dimension ", dim, " instead."
   );
-  auto self_ptr = get_nested_tensor_impl(self);
-  // buffer contains the underlying data in a contiguous vector
-  const at::Tensor & buffer = self_ptr->get_buffer();
-  int64_t numel = buffer.numel();
-  TORCH_CHECK(
-    numel > 0,
-    "cannot index an empty nested tensor."
-  );
-  // nested_tensor[i] = i-th original tensor
-  int64_t ntensors = *(self_ptr->opt_size(0));
-  int64_t positive_index = at::maybe_wrap_dim(index, ntensors);
-  // determine the memory segment of the i-th original tensor
-  Tensor sizemat = get_nested_size_tensor(self);
-  int64_t original_dim = sizemat.size(1);
-  const int64_t * sizemat_ptr = sizemat.data_ptr<int64_t>();
-  // start of the segment
-  int64_t start = 0, sizemat_offset = 0;
-  for (int64_t i = 0; i < positive_index; i++) {
-    int64_t row_product = sizemat_ptr[sizemat_offset];
-    sizemat_offset++;
-    for (int64_t j = 1; j < original_dim; j++) {
-      row_product *= sizemat_ptr[sizemat_offset];
-      sizemat_offset++;
-    }
-    start += row_product;
-  }
-  // btw determine the shape of the i-th original tensor
-  IntArrayRef shape(sizemat_ptr + sizemat_offset, sizemat_ptr + sizemat_offset + original_dim);
-  // stop of the segment
-  int64_t stop;
-  if (positive_index == ntensors - 1) {
-    stop = numel;
+  int64_t ntensors = 0;
+  if (self_ptr->get_nested_size_tensor().dim() == 2) {
```

Comment:
The check against 2 here is a bit strange even if correct; maybe `self_ptr->get_nested_size_tensor().dim() > 0` is nicer?

Reply:
Same feeling here, but the ==2 check is what people have done so I just followed this practice.

From what I understand, the size tensor can have 2 possible `dim()`:
* 0, when the nested tensor is empty
* 2, otherwise

Reply:
added `NestedTensorImpl::size` as suggested
```
  int64_t size(int64_t d) const {
    c10::optional<int64_t> optional_size = this->opt_size(d);
    TORCH_CHECK(optional_size.has_value(),
        "Given dimension ", d, " is irregular and does not have a size.");
    return *optional_size;
  }
```